pkg/relayer/session/session.go

```diff
@@ -146,12 +146,19 @@ func (rs *relayerSessionsManager) mapBlockToSessionsToClaim(
 	_ context.Context,
 	block client.Block,
 ) (sessionTrees []relayer.SessionTree, skip bool) {
+	rs.sessionsTreesMu.Lock()
+	defer rs.sessionsTreesMu.Unlock()
+
 	// Check if there are sessions that need to enter the claim/proof phase
 	// as their end block height was the one before the last committed block.
 	// Iterate over the sessionsTrees map to get the ones that end at a block height
 	// lower than the current block height.
 	for endBlockHeight, sessionsTreesEndingAtBlockHeight := range rs.sessionsTrees {
-		if endBlockHeight < block.Height() {
+		// TODO: We need this to be == instead of <= because we don't want to keep sending
```

Comment:
@Olshansk , Definitely a tech debt. Completely resolving this is kinda delicate, since it involves adding persistence to check proofs after the service is restarted in addition to a retry mechanism.

I'd suggest to do the following:
a. Runtime checks:
  1.  Add a `submitting` or `waiting` property to the `SessionTree` struct to let this function skip `SessionTree`s that are already being handled
  2. Have a retry mechanism on the `createClaim` that handles failing submissions
b: Persistence (Boot time checks)
  1. Persist the last successful claim submission
  2. Fast forward and collect all sessions til the latest one and try to handle them (see if there is any work done, if so submit them)
  3. Do the same for proofs

cc @bryanchriswhite 

If you agree that this is the right solution, I'll create a proper ticket for it and update the TODO.

Reply:
> a. Runtime checks:
>  1. Add a submitting or waiting property to the SessionTree struct to let this function skip SessionTrees that are already being handled

:+1: wdyt about `pending bool`?

> 2. Have a retry mechanism on the createClaim that handles failing submissions

I would propose adding new unexported methods/mapFns on `relayerSessionsManager` which take the `failedSession...` observables and would be called in the same scope as `#createClaims()` add `#submitProofs()` methods.

> b. Persistence (Boot time checks)
>   1. Persist the last successful claim submission
> Fast forward and collect all sessions til the latest one and try to handle them (see if there is any work done, if so submit them)
> Do the same for proofs

:+1:

Reply:
> I would propose adding new unexported methods/mapFns on relayerSessionsManager which take the failedSession... observables and would be called in the same scope as #createClaims() add #submitProofs() methods.

This :point_up_2:

Reply:
Adding a link to this discussion in the comments with a `TODO_BLOCKER`